### DISCUSSION TITLE
Can I use a GtkDrawingArea in a widget ui file?

### DISCUSSION THREAD

**Discussion by avrame** [hugopl/gtk4.cr#66] [Q&A]
I'm experimenting with making a simple web browser using Crystal and Gtk4 (following [this guide](https://limpet.net/mbrubeck/2014/08/08/toy-layout-engine-1.html))

I modified the widget template example to add a `GtkDrawingArea` by adding this: 
```
<child>
  <object class="GtkDrawingArea" id="canvas" />
</child>
```

And then in my crystal file I added this:
```
@canvas = Gtk::DrawingArea.cast(template_child("canvas"))
```

So far, so good - it compiles, runs and still displays the window. But from reading the Gtk4 docs on [GtkDrawingArea](https://docs.gtk.org/gtk4/class.DrawingArea.html) I can't figure out how to add the drawing function in Crystal (I don't really know C)

**Comment by hugopl** [comment]
You need to be notified when the widget was resized, for this you can use the `resize` signal or create a `resize` virtual function if you are creating a class that inherits `Gtk::DrawingArea`.

To draw things you need to call [Gtk::DrawingArea#draw_fund](https://hugopl.github.io/gtk4.cr/Gtk/DrawingArea.html#draw_func%3D%28draw_func%3AGtk%3A%3ADrawingAreaDrawFunc%7CNil%29%3ANil-instance-method).

It receives a [Gtk::DrawingAreaDrawFunc](https://hugopl.github.io/gtk4.cr/Gtk/DrawingAreaDrawFunc.html) that is just a alias to a proc `Gtk::DrawingArea, Cairo::Context, Int32, Int32 -> Nil`

so...

```Crystal
class MyCuteDrawingArea < Gtk::DrawingArea
  def initialize
    super() # we must call the constructor of base class to initialize gobjet stuff
    self.draw_func = ->draw(Gtk::DrawingArea, Cairo::Context, Int32, Int32)
  end

  private def draw(area, context, width, height) : Nil
    puts "Hey mon! I'm (almost) drawing!!"
  end

  @[GObject::Virtual]
  def resize(width : Int32, height : Int32) : Nil
    puts "widget resized"
  end
end
```

BTW, if you are not doing a pixel by pixel drawing may be worth to check GSK, the "new" way of paint widgets in GTK4, see [examples/particles.cr](https://github.com/hugopl/gtk4.cr/blob/main/examples/particles.cr#L61) for an example.

I used GTKDrawingArea in GTK3 but never tested it with these GTK4 bindings, good luck and report any issues.

**Comment by hugopl** [comment]
Ah, answering your question, yes, you can use Gtk::DrawingArea from UI too, so you need to listen to the `resize` signal since you can't implement the virtual function.

**Reply by avrame** [reply]
Thanks for your help! Are there any advantages to using GSK instead of GtkDrawingArea?

**Reply by hugopl** [reply]
GSK is just a new rendering engine API, done in order to render the widget using modern features like hardware accelerated rendering.

GTK itself moved to GSK, so I would follow the herd. However until some versions ago wasn't possible to render more complex shapes like paths, cairo was used for those cases. So use GSK if you can.

**Reply by avrame** [reply]
Ok, GSK sounds like the way to go! Hey, I tried compiling this code:
```crystal
require "gtk4"

module UI
  class CanvasWidget < Gtk::Widget
    gsk_rect : Gsk::RoundedRect

    def initialize
      super
      position = Graphene::Point.new(10, 10)
      dimensions = Graphene::Size.new(20, 40)
      graphene_rect = Graphene::Rect.new(position, dimensions)
      @gsk_rect = Gsk::RoundedRect.new(graphene_rect)
    end

    @[GObject::Virtual]
    def snapshot(snapshot : Gtk::Snapshot)
      # snapshot.append_border
    end
  end

  class ChrysalisApp < Gtk::Application
    def initialize
      super(application_id: "chrysalis.com", flags: Gio::ApplicationFlags::None)
    end

    @[GObject::Virtual]
    def activate
      window = Gtk::ApplicationWindow.new(self)
      window.title = "Chrysalis Browser"
      window.set_default_size(640, 480)
      window.child = CanvasWidget.new
      window.present
    end
  end

  exit(ChrysalisApp.new.run)
end
```
But the compiler outputs this error:
```console
Error: instantiating 'UI::ChrysalisApp#activate()'


In src/ui.cr:31:35

 31 | window.child = CanvasWidget.new
                                  ^--
Error: instantiating 'UI::CanvasWidget.new()'


In src/ui.cr:11:38

 11 | graphene_rect = Graphene::Rect.new(position, dimensions)
                                     ^--
Error: instantiating 'Graphene::Rect.new(Graphene::Point, Graphene::Size)'


In lib/gi-crystal/src/auto/graphene-1.0/rect.cr:18:13

 18 | @data.origin = origin unless origin.nil?
            ^-----
Error: instantiating 'LibGraphene::Rect#origin=(Graphene::Point)'


In lib/gi-crystal/src/auto/graphene-1.0/lib_graphene.cr:66:5

 66 | origin : LibGraphene::Point
      ^
Error: invoked 'to_unsafe' to convert from Graphene::Point to LibGraphene::Point, but got Pointer(Void)
avrameisner@AvramsMBPM114 chrysalis % 
```

**Reply by avrame** [reply]
@hugopl Should I submit an issue for the above problem that I'm having compiling the code or is there something wrong with my code?

**Reply by hugopl** [reply]
Your code seems fine, it's a bug, the constructor code should be something like:

```Crystal
module Graphene
  struct Rect
    def initialize(origin : Graphene::Point? = nil, size : Graphene::Size? = nil)
      @data = LibGraphene::Rect.new
      @data.origin = origin.as(Pointer(LibGraphene::Point)).value unless origin.nil?
      @data.size = size.as(Pointer(LibGraphene::Size)).value unless size.nil?
    end
  end
end
```

Structs with member structs that are both binded as crystal structs generate this bad code at constructor.

You can try to fix that, the issue is at `lib/gi-crystal/src/generator/stack_struct_gen.cr:25`, the method `StackStructGen#render_initialize`.

To known the side effects of your change you can call `make compare`, then later modify things and call `make recompare`, if you are on Linux and have `meld` installed it generate the test bindings then opens meld to show the diff between old and new version.

**Reply by avram-at-hazel** [reply]
Thanks, I'll try that tonight!

**Reply by hugopl** [reply]
`make test` build everything and run the tests, if you write some test and just want to re-run it without re-generate the bindings you can do `./bin/spec`, it sets the necessary environment variables so the tests can find the C library (libtest) used as test binding.
